R/zchunk_L221.en_supply.R

```diff
@@ -92,295 +105,661 @@ module_energy_L221.en_supply_DISABLED <- function(command, ...) {
+                           GCAM_region_names = GCAM_region_names) -> L221.SubsectorLogit_en
+
+    # Subsector shareweights of upstream energy handling sectors
+    if(any(!is.na(A21.subsector_shrwt$year))){
+      A21.subsector_shrwt %>%
+        write_to_all_regions(c(LEVEL2_DATA_NAMES[["SubsectorShrwt"]]), has_traded = TRUE,
+                             GCAM_region_names = GCAM_region_names) %>%
+        filter(!is.na(year)) -> L221.SubsectorShrwt_en
+    }
+
+    if(any(!is.na(A21.subsector_shrwt$year.fillout))){
+      A21.subsector_shrwt %>%
+        write_to_all_regions(c(LEVEL2_DATA_NAMES[["SubsectorShrwtFllt"]]), has_traded = TRUE,
+                             GCAM_region_names = GCAM_region_names) %>%
+        filter(!is.na(year.fillout)) -> L221.SubsectorShrwtFllt_en
+    }
+
+    # Subsector shareweight interpolation of upstream energy handling sectors
+    if(any(is.na(A21.subsector_interp$to.value))){
+      A21.subsector_interp %>%
+        write_to_all_regions(c(LEVEL2_DATA_NAMES[["SubsectorInterp"]], "to.value"), has_traded = TRUE,
+                             GCAM_region_names = GCAM_region_names) %>%
+        filter(is.na(to.value)) %>%
+        select(-to.value) -> L221.SubsectorInterp_en
+    }
+
+    if(any(!is.na(A21.subsector_interp$to.value))){
+      A21.subsector_interp %>%
+        write_to_all_regions(c(LEVEL2_DATA_NAMES[["SubsectorInterpTo"]]), has_traded = TRUE,
+                             GCAM_region_names = GCAM_region_names) %>%
+        filter(!is.na(to.value)) -> L221.SubsectorInterpTo_en
+    }
+
+    # Technology information
+    # Identification of stub technologies (assume those in global tech shareweight table include all techs)
+    A21.globaltech_shrwt %>%
+      write_to_all_regions(c(LEVEL2_DATA_NAMES[["Tech"]]), has_traded = FALSE,
+                           GCAM_region_names = GCAM_region_names) -> L221.StubTech_en
+
+    #Drop stub technologies for biomassOil techs that do not exist
+    A21.globaltech_shrwt %>%
+      filter(supplysector == "regional biomassOil") %>%
+      select(supplysector, subsector, technology) %>%
+      repeat_add_columns(tibble("GCAM_region_ID" = GCAM_region_names$GCAM_region_ID, "region" = GCAM_region_names$region)) %>%
+      left_join(A_regions %>%
+                  rename(technology = biomassOil_tech), by = c("region", "GCAM_region_ID", "technology")) %>%
+      filter(is.na(tradbio_region)) %>%
+      select(supplysector, subsector, technology, region, GCAM_region_ID) -> L221.rm_biomassOil_techs_R
+
+    L221.StubTech_en %>%
+      anti_join(L221.rm_biomassOil_techs_R, by = c("region", "technology")) %>%
+      rename(stub.technology = technology) -> L221.StubTech_en
+
+    # Coefficients of global technologies
+    # Energy inputs and coefficients of global technologies for upstream energy handling
+
+    A21.globaltech_coef %>%
+      gather(year, coef, matches(YEAR_PATTERN)) %>%
+      mutate(year = as.integer(year)) -> A21.globaltech_coef
```

Comment:
For these two lines, and any similar others below, please use new `gather_years` function (see many examples throughout code: 201-202, 218-219, etc).